tests/utils/test_parameterized_query.py

```diff
@@ -131,3 +158,21 @@ def test_raises_on_unexpected_param_types(self):
 
         with pytest.raises(InvalidParameterError):
             query.apply({"bar": "baz"})
+
+    @patch('redash.models.QueryResult.get_by_id_and_org', return_value=namedtuple('query_result', 'data')('{"rows": [{"id": 5, "name": "John", "value": "John Doe"}]}'))
+    @patch('redash.utils.parameterized_query.require_access')
+    @patch('redash.utils.parameterized_query.view_only')
+    @patch('redash.utils.parameterized_query.current_user')
+    @patch('redash.models.Query')
+    def test_dropdown_values_prefers_name_and_value_columns(self, *_):
```

Comment:
And here's why moving that pluck function to the module level would've been beneficial: you could've tested it directly without all these mocks.

Reply:
(I know you prefer to test the external API, but in this case it feels just too much)

Reply:
I agree with the general idea here of not testing implementation details. But maybe this function should be two functions composed together when used?

```python
def load_query_latest_result(query_id):
    ... magic ...

def dropdown_values_from_query_result(query_result):
    ... darker magic ...

# then in the dropdown handler:
class QueryDropDownResource(BaseResource):
    def get(query_id):
          # QueryResult#json is a new property that will return data post json_load
          return dropdown_values_from_query_result(load_query_latest_result(query_id).json)
```

Each function on its own has a clear purpose. You can say that "extracting dropdown values" is an implementation detail that should be hidden, because that's the whole thing here. So having it in its own function won't slow down changes. If it all, it will make them easier, because you don't need to worry about all the aspects of loading the query result.

Then you need the heavy mocking only once, in the integration test for `QueryDropDownResource`.

Reply:
*** cracks knuckles ***

Yes, separation of concerns would be beneficial here. I definitely think we should extract all the data loading to a helper function and then `dropdown_values` would be basically:

```
rows = _load_result(query_id, org)
return map(_pluck_name_and_value, rows)
```

Regarding testing of pluck logic - when I find myself wanting to test private stuff, my method has always been asking myself “is this such an important piece of logic that I have to test it directly?”. If the answer is yes, it means that the test subject is _too important_ to be hidden inside something else - it is probably a domain concept that needs to be promoted to its own place.

In this case, I think that your motivation for promoting plucking logic is not due to importance (dropdown values is a really small concept for Redash), but due to the fuglyness caused by multiple calls to `patch`.

I’m going to guess you will find 739c65e3 more pleasing.

Reply:
It looks like we're in an agreement here on the concepts. It's just the details we're seeing in a different way. It's not about how ugly the tests are (we have plenty of those...). They were just the signal.

I'm not saying the tests should test the pluck logic, I'm saying the tests should test the "convert query results to dropdown values" logic. Pluck is just an implementation detail that is definitely not interesting on its own.

Also, the query result loading logic is very important as well (data security) and should be tested as well...

Reply:
Anyway, up to you how to handle this, but the important thing is to test the security aspects.